File: ui/hub-mk3-ui/src/store/orm/program-voice.js
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.
import {attr, fk, Model} from "redux-orm";

class ProgramVoice extends Model {
  toString() {
    return `ProgramVoice: ${this.name}`;
  }

  // Declare any static or instance methods you need.
}

ProgramVoice.modelName = 'ProgramVoice';

// Declare your related fields.
ProgramVoice.fields = {
  id: attr(),
  density: attr(),
  key: attr(),
  name: attr(),
  tempo: attr(),
  programId: fk({
    to: 'Program',
    as: 'program',
    relatedName: 'voices',
  }),
  /*
  segmentChoiceArrangements: many('segment-choice-arrangement'),
  programSequencePatterns: many('program-sequence-pattern'),
   */
};

export default ProgramVoice;

File: engine/src/content/ProgramVoice.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <set>
#include "xjmusic/content/ProgramVoice.h"

std::set<std::string> XJ::ProgramVoice::getNames(const std::set<const ProgramVoice *>& voices) {
  std::set<std::string> names;
  for (const ProgramVoice *voice : voices) {
    names.insert(voice->name);
  }
  return names;
}
